.circleci/meta-linter/main.go

```diff
@@ -18,21 +18,104 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"bitbucket.org/creachadair/stringset"
 )
 
 func main() {
 	errors := []error{}
 
-	// Run each checker and run up the list of errors.
-	errors = append(errors, checkRulesRegistered()...)
-	errors = append(errors, ruleFilename()...)
-	errors = append(errors, checkRulesDocumented()...)
+	// Keep track of rules processed, and which ones have one or more failures.
+	pass := stringset.New()
+	fail := stringset.New()
+
+	// Begin walking the rules directory looking for rules.
+	err := filepath.Walk("./rules", func(path string, info os.FileInfo, err error) error {
+		// Sanity check: Bubble errors.
+		if err != nil {
+			errors = append(errors, err)
+			return nil
+		}
+
+		// Weed out files that are not rules (tests, etc.).
+		for _, exempt := range []func(path string, info os.FileInfo) bool{
+			func(_ string, i os.FileInfo) bool { return i.IsDir() },
+			func(p string, _ os.FileInfo) bool { return strings.Contains(p, "/internal/") },
+			func(p string, _ os.FileInfo) bool { return p == "rules/rules.go" },
+			func(p string, _ os.FileInfo) bool { return strings.HasSuffix(p, "_test.go") },
+			func(p string, _ os.FileInfo) bool { return aipIndex.MatchString(p) },
+		} {
+			if exempt(path, info) {
+				return nil
+			}
+		}
+
+		// This represents a rule. Get the AIP and rule name from the path.
+		match := ruleFile.FindStringSubmatch(path)
```

Comment:
Check if `len(match) == 3`.

Reply:
That feels like a change detector condition. The length of the match is guaranteed to be 3 if there is a match at all.

Reply:
Fair enough. Thanks!